drivers/flash/nrf_qspi_nor.c

```diff
@@ -493,14 +493,14 @@ static int qspi_nor_read(struct device *dev, off_t addr, void *dest,
 {
 	void *dptr = dest;
 	size_t dlen = size;
-	u8_t buf[4];
+	u8_t __aligned(4) buf[4];
 
 	if (!dest) {
 		return -EINVAL;
 	}
 
 	/* read size must be non-zero multiple of 4 bytes */
-	if (size < 4U) {
+	if ((size > 0) && (size < 4U)) {
```

Comment:
Since you are digging here anyway, maybe just gather all the input parameter checks (`addr`, `dest`, `size`)that end up in `-EINVAL` to line 498?

Reply:
The condition becomes:
```
	if ((dest == NULL)
	    || (size == 0)
	    || ((size > 4U) && ((size % 4U) != 0))
	    || ((addr % 4U) != 0)
	    || (addr >= params->size)
	    || (addr < 0)
	    || (size > params->size)
	    || (((addr) + size) > params->size)) {
		return -EINVAL;
	}
```
assuming we don't eliminate the redundant checks but do eliminate the error message for address violations (why is that more important to diagnose than others?).

I can see some merit in that approach, but also merit for checking each parameter separately where we can document the reasons for the check and (where appropriate) issue specific diagnostics.

At this point in the release cycle I prefer to do minimal, more clearly-correct changes that retain consistency with the other functions code.

Reply:
I was rather obvious checks in mind, like parameter sanity check, these you do not log error for, just return.
```
	if ((dest == NULL)
	    || (size == 0)
	    || ((size > 4U) && ((size % 4U) != 0))
	    || ((addr % 4U) != 0)
}
```
This way we would not do anything that would get discarded anyway on next check (I know that there is not much done here).

> I can see some merit in that approach, but also merit for checking each parameter separately where we can document the reasons for the check and (where appropriate) issue specific diagnostics.

I get your point.

>At this point in the release cycle I prefer to do minimal, more clearly-correct changes that retain consistency with the other functions.

Have not thought about this, my bad.